cpp/src/experimental/louvain.cuh

```diff
@@ -561,58 +188,94 @@ class Louvain {
+        thrust::make_tuple(cluster_keys_v_.begin(), cluster_weights_v_.begin()));
+
+      std::forward_as_tuple(std::tie(rx_keys_v, rx_weights_v), std::ignore) =
+        groupby_gpuid_and_shuffle_values(
+          handle_.get_comms(),
+          pair_first,
+          pair_first + current_graph_view_.get_number_of_local_vertices(),
+          [key_func =
+             cugraph::experimental::detail::compute_gpu_id_from_vertex_t<vertex_t>{
+               comm_size}] __device__(auto val) { return key_func(thrust::get<0>(val)); },
+          handle_.get_stream());
+
+      cluster_keys_v_    = std::move(rx_keys_v);
+      cluster_weights_v_ = std::move(rx_weights_v);
+    }
 
     timer_stop(stream_);
   }
 
-  template <typename iterator_t, typename T>
-  void cache_vertex_properties(iterator_t const &local_input_iterator,
-                               rmm::device_vector<T> &src_cache_v,
-                               rmm::device_vector<T> &dst_cache_v,
-                               bool src = true,
-                               bool dst = true)
+  template <typename T>
+  T *cache_src_vertex_properties(rmm::device_uvector<T> &input, rmm::device_uvector<T> &src_cache_v)
   {
-    if (src) {
-      src_cache_v.resize(current_graph_view_.get_number_of_local_adj_matrix_partition_rows());
-      copy_to_adj_matrix_row(
-        handle_, current_graph_view_, local_input_iterator, src_cache_v.begin());
+    if (graph_view_t::is_multi_gpu) {
+      src_cache_v.resize(current_graph_view_.get_number_of_local_adj_matrix_partition_rows(),
+                         stream_);
+      copy_to_adj_matrix_row(handle_, current_graph_view_, input.begin(), src_cache_v.begin());
+      return src_cache_v.begin();
+    } else {
+      return input.begin();
     }
+  }
 
-    if (dst) {
-      dst_cache_v.resize(current_graph_view_.get_number_of_local_adj_matrix_partition_cols());
-      copy_to_adj_matrix_col(
-        handle_, current_graph_view_, local_input_iterator, dst_cache_v.begin());
+  template <typename T>
+  T *cache_dst_vertex_properties(rmm::device_uvector<T> &input, rmm::device_uvector<T> &dst_cache_v)
+  {
+    if (graph_view_t::is_multi_gpu) {
+      dst_cache_v.resize(current_graph_view_.get_number_of_local_adj_matrix_partition_cols(),
+                         stream_);
+      copy_to_adj_matrix_col(handle_, current_graph_view_, input.begin(), dst_cache_v.begin());
+      return dst_cache_v.begin();
+    } else {
+      return input.begin();
     }
   }
 
-#ifdef CUCO_STATIC_MAP_DEFINED
+  template <typename T>
+  std::tuple<T *, T *> cache_vertex_properties(rmm::device_uvector<T> &input,
+                                               rmm::device_uvector<T> &src_cache_v,
+                                               rmm::device_uvector<T> &dst_cache_v)
+  {
+    auto src = cache_src_vertex_properties(input, src_cache_v);
+    auto dst = cache_dst_vertex_properties(input, dst_cache_v);
+
+    return std::make_tuple(src, dst);
+  }
```

Comment:
Do we really need this function anymore? This may cut one line in the caller site but just add an additional layer of indirection to track.

Reply:
Deleted this.  In my original MNMG implementation this was called dozens of times.